pkg/gameserversets/controller.go

```diff
@@ -0,0 +1,345 @@
+	list, err := c.gameServerLister.List(labels.SelectorFromSet(labels.Set{stablev1alpha1.GameServerSetGameServerLabel: gss.ObjectMeta.Name}))
+	if err != nil {
+		return list, errors.Wrapf(err, "error listing gameservers for gameserverset %s", gss.ObjectMeta.Name)
+	}
+
+	var result []*stablev1alpha1.GameServer
+	for _, gs := range list {
+		if metav1.IsControlledBy(gs, gss) {
+			result = append(result, gs)
+		}
+	}
+
+	return result, nil
+}
+
+// syncUnhealthyGameServers deletes any unhealthy game servers (that are not already being deleted)
+func (c *Controller) syncUnhealthyGameServers(gss *stablev1alpha1.GameServerSet, list []*stablev1alpha1.GameServer) error {
+	for _, gs := range list {
+		if gs.Status.State == stablev1alpha1.Unhealthy && gs.ObjectMeta.DeletionTimestamp.IsZero() {
+			err := c.gameServerGetter.GameServers(gs.ObjectMeta.Namespace).Delete(gs.ObjectMeta.Name, nil)
+			if err != nil {
+				return errors.Wrapf(err, "error deleting gameserver %s", gs.ObjectMeta.Name)
+			}
+			c.recorder.Eventf(gss, corev1.EventTypeNormal, "UnhealthyDelete", "Deleted gameserver: %s", gs.ObjectMeta.Name)
+		}
+	}
+
+	return nil
+}
+
+// syncMoreGameServers adds diff more GameServers to the set
+func (c *Controller) syncMoreGameServers(gss *stablev1alpha1.GameServerSet, diff int32) error {
+	if diff <= 0 {
+		return nil
+	}
+	c.logger.WithField("diff", diff).WithField("gameserverset", gss.ObjectMeta.Name).Info("Adding more gameservers")
+	for i := int32(0); i < diff; i++ {
+		gs := gss.GameServer()
+		gs, err := c.gameServerGetter.GameServers(gs.Namespace).Create(gs)
+		if err != nil {
+			return errors.Wrapf(err, "error creating gameserver for gameserverset %s", gss.ObjectMeta.Name)
+		}
+		c.recorder.Eventf(gss, corev1.EventTypeNormal, "SuccessfulCreate", "Created gameserver: %s", gs.ObjectMeta.Name)
+	}
+
+	return nil
+}
+
+// syncLessGameSevers removes Ready GameServers from the set of GameServers
+func (c *Controller) syncLessGameSevers(gss *stablev1alpha1.GameServerSet, list []*stablev1alpha1.GameServer, diff int32) error {
+	if diff >= 0 {
+		return nil
+	}
+	// easier to manage positive numbers
+	diff = -diff
+	c.logger.WithField("diff", diff).WithField("gameserverset", gss.ObjectMeta.Name).Info("Deleting gameservers")
+	count := int32(0)
+
+	// count anything that is already being deleted
+	for _, gs := range list {
+		if !gs.ObjectMeta.DeletionTimestamp.IsZero() {
+			diff--
+		}
+	}
+
+	for _, gs := range list {
+		if diff <= count {
+			return nil
+		}
+
+		if gs.Status.State != stablev1alpha1.Allocated {
+			err := c.gameServerGetter.GameServers(gs.Namespace).Delete(gs.ObjectMeta.Name, nil)
+			if err != nil {
+				return errors.Wrapf(err, "error deleting gameserver for gameserverset %s", gss.ObjectMeta.Name)
+			}
+			c.recorder.Eventf(gss, corev1.EventTypeNormal, "SuccessfulDelete", "Deleted GameServer: %s", gs.ObjectMeta.Name)
+			count++
+		}
+	}
```

Comment:
What happens if they are all allocated? Will this re-queue the sync event to be processed later on?

Reply:
If they are allocated, [the event handler above](https://github.com/GoogleCloudPlatform/agones/pull/156/commits/ef92839c3ac99bdf224ec0d16adf367d40098bc1#diff-3d3de0ef142666d0fe31e84b75d29d9cR119) will watch for when the `GameServer` is deleted for this `GameServerSet`, and when it catches that event, it will queue the owning `GameServerSet` to be processed again, and make sure it's at the right number.